File: backend/routers/validator.py
"""
Validator router — clone a participant's GitHub repo and run section tests against it.

Rules:
  - Each section has a separate time window set by admin.
  - After the window expires, validation is rejected.
  - Multiple submissions allowed; only the highest score is kept (DB GREATEST logic).
"""

import subprocess
import json
import os
import time
import tempfile
import shutil
import resource
from datetime import datetime, timezone, timedelta

from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel, Field
from typing import List, Optional

from core.config import settings
from core.deps import get_current_user
from core.database import (
    upsert_score,
    get_section_timer,
    save_anti_cheat_report,
    find_duplicate_submission_hash,
    count_recent_submissions,
    log_editor_activity,
    get_editor_activity_metrics,
)
from anti_cheat import verify_test_integrity, scan_suspicious_imports, hash_source_tree
from risk_engine import build_risk

router = APIRouter()

# Points per section (max)
SECTION_POINTS = {1: 100, 2: 100, 3: 100, 4: 120}

# Relative paths to test directories inside the cloned repo
SECTION_TESTS = {
    1: "section1-multifile-debug/tests",
    2: "section2-broken-recovery/tests",
    3: "section3-memory-deadlock/tests",
    4: "section4-logical-tracing/tests",
}


class ValidateRequest(BaseModel):
    section: int
    github_url: str


class EditorActivityRequest(BaseModel):
    event: str
    team_id: Optional[str] = None
    timestamp: Optional[str] = None
    editor_length_before: int = 0
    editor_length_after: int = 0
    chars_delta: int = 0
    delta_ms: int = 0
    typing_speed_cps: float = 0.0
    flags: List[str] = Field(default_factory=list)
    details: dict = Field(default_factory=dict)


@router.post("/editor-activity")
def track_editor_activity(body: EditorActivityRequest, user: dict = Depends(get_current_user)):
    """Receive client-side editor telemetry events (paste attempts / typing anomalies)."""
    team_id = body.team_id or user.get("team") or user.get("sub") or "unknown"
    metadata = {
        "timestamp": body.timestamp,
        "flags": body.flags,
        "details": body.details,
    }
    log_id = log_editor_activity(
        username=user.get("sub", ""),
        team_id=team_id,
        event=body.event,
        editor_length_before=body.editor_length_before,
        editor_length_after=body.editor_length_after,
        chars_delta=body.chars_delta,
        delta_ms=body.delta_ms,
        typing_speed_cps=body.typing_speed_cps,
        metadata=metadata,
    )
    return {"ok": True, "id": log_id}


def _limit_process_resources() -> None:
    """Child-process hardening: memory cap + soft CPU limit."""
    mem_limit_mb = int(os.getenv("PYTEST_MEMORY_LIMIT_MB", "512"))
    mem_bytes = mem_limit_mb * 1024 * 1024

    resource.setrlimit(resource.RLIMIT_AS, (mem_bytes, mem_bytes))
    cpu_cap = max(settings.PYTEST_TIMEOUT + 5, 10)
    resource.setrlimit(resource.RLIMIT_CPU, (cpu_cap, cpu_cap))


@router.post("/section")
def validate_section(body: ValidateRequest, user: dict = Depends(get_current_user)):
    """
    Clone participant's GitHub repo and run pytest for the requested section.
    Returns pass/fail breakdown and updates the leaderboard (highest score kept).
    """
    if body.section not in SECTION_TESTS:
        raise HTTPException(status_code=422, detail="section must be 1–4")

    # ── Timer check ───────────────────────────────────────────────────────────
    timer = get_section_timer(body.section)
    if timer and timer.get("is_active") and timer.get("start_time"):
        start    = timer["start_time"]
        elapsed  = timer.get("elapsed_seconds") or 0
        paused_at = timer.get("paused_at")

        def _tz(dt):
            if dt and hasattr(dt, "utcoffset") and dt.utcoffset() is None:
                return dt.replace(tzinfo=timezone.utc)
            return dt

        start     = _tz(start)
        paused_at = _tz(paused_at)
        total_secs = timer["duration_minutes"] * 60
        now = datetime.now(timezone.utc)

        if paused_at:
            # Timer is paused — remaining = total - elapsed_at_pause
            remaining = total_secs - elapsed
        else:
            # Timer is running
            running_secs = int((now - start).total_seconds())
            remaining = total_secs - elapsed - running_secs

        if remaining <= 0:
            raise HTTPException(
                status_code=403,
                detail=f"Section {body.section} time has expired. No more submissions accepted."
            )

    # ── Clone repo ────────────────────────────────────────────────────────────
    tmpdir = tempfile.mkdtemp(prefix="dc_validate_")
    report_path = None
    raw_stdout = ""
    raw_stderr = ""
    test_hash_valid = True
    test_hash_mismatches = []
    suspicious_imports = []
    suspicious_occurrences = 0
    submission_hash = ""
    duplicate_detected = False
    duplicate_with = None
    submission_rate_last_10min = 0
    copy_metrics = {
        "paste_attempts": 0,
        "large_injection_events": 0,
        "typing_anomaly_detected": False,
        "typing_anomaly_events": 0,
        "average_typing_speed_cps": 0.0,
        "copy_risk_score": 0,
        "tab_switches": 0,
        "window_blur_seconds": 0.0,
    }
    child_cpu_before = resource.getrusage(resource.RUSAGE_CHILDREN)
    child_mem_before_kb = child_cpu_before.ru_maxrss
    try:
        # Disable all credential helpers so git never hangs waiting for
        # a username/password prompt — only works with public repos.
        clone_env = {
            **os.environ,
            "GIT_TERMINAL_PROMPT": "0",
            "GIT_ASKPASS": "echo",
        }
        clone = subprocess.run(
            [
                "git", "clone",
                "-c", "credential.helper=",
                "--depth=1", "--no-tags",
                body.github_url, tmpdir,
            ],
            capture_output=True, text=True, timeout=90,
            env=clone_env,
        )
        if clone.returncode != 0:
            raise HTTPException(
                status_code=422,
                detail=f"Failed to clone repository: {clone.stderr[:400].strip()}"
            )

        # ── Verify tests exist ────────────────────────────────────────────────
        test_path = os.path.join(tmpdir, SECTION_TESTS[body.section])
        if not os.path.isdir(test_path):
            raise HTTPException(
                status_code=422,
                detail=f"Tests for section {body.section} not found in repo. "
                       f"Expected: {SECTION_TESTS[body.section]}"
            )

        # ── Anti-cheat pre-checks ───────────────────────────────────────────
        team_id = user.get("team") or user.get("sub") or "unknown"
        test_hash_valid, test_hash_mismatches = verify_test_integrity(settings.REPO_ROOT, tmpdir)
        suspicious_imports, suspicious_occurrences = scan_suspicious_imports(tmpdir)
        submission_hash = hash_source_tree(tmpdir)

        duplicate_row = find_duplicate_submission_hash(submission_hash, team_id)
        duplicate_detected = duplicate_row is not None
        duplicate_with = duplicate_row.get("team_id") if duplicate_row else None

        submission_rate_last_10min = count_recent_submissions(team_id=team_id, minutes=10)
        copy_metrics = get_editor_activity_metrics(team_id=team_id, minutes=10)

        # ── Run pytest ────────────────────────────────────────────────────────
        with tempfile.NamedTemporaryFile(suffix=".json", delete=False) as f:
            report_path = f.name

        start_t = time.monotonic()
        run_result = subprocess.run(
            [
                "python", "-m", "pytest", test_path,
                "--json-report", f"--json-report-file={report_path}",
                "--tb=no", "-q",
                f"--timeout={settings.PYTEST_TIMEOUT}",
            ],
            capture_output=True, text=True,
            timeout=settings.PYTEST_TIMEOUT + 30,
            cwd=tmpdir,
            preexec_fn=_limit_process_resources,
        )
        duration = time.monotonic() - start_t
        raw_stdout = run_result.stdout or ""
        raw_stderr = run_result.stderr or ""

    finally:
        shutil.rmtree(tmpdir, ignore_errors=True)

    # ── Parse report ──────────────────────────────────────────────────────────
    try:
        with open(report_path) as f:
            report = json.load(f)
        os.unlink(report_path)
    except Exception:
        raise HTTPException(status_code=500, detail="Could not parse pytest report")

    summary = report.get("summary", {})
    passed  = summary.get("passed", 0)
    failed  = summary.get("failed", 0) + summary.get("error", 0)
    total   = summary.get("total", passed + failed)
    skipped = summary.get("skipped", 0)
    collected = summary.get("collected", total)

    child_cpu_after = resource.getrusage(resource.RUSAGE_CHILDREN)
    cpu_time_delta = (
        (child_cpu_after.ru_utime + child_cpu_after.ru_stime)
        - (child_cpu_before.ru_utime + child_cpu_before.ru_stime)
    )
    cpu_usage_percent = round((cpu_time_delta / duration) * 100, 2) if duration > 0 else 0.0
    mem_after_kb = child_cpu_after.ru_maxrss
    mem_delta_kb = max(mem_after_kb - child_mem_before_kb, 0)
    memory_usage_mb = round(mem_delta_kb / 1024.0, 2)

    max_pts = SECTION_POINTS[body.section]
    score   = round((passed / total * max_pts) if total else 0, 2)

    # Parse results
    def _get_test_error(t):
        for phase in ("call", "setup", "teardown"):
            pd = t.get(phase) or {}
            if pd.get("outcome") in ("failed", "error"):
                lr = pd.get("longrepr", "")
                if isinstance(lr, dict):
                    lr = lr.get("repr", "")
                return str(lr).strip()
        return ""

    tests = [
        {
            "name":     t.get("nodeid", "").split("::")[-1],
            "outcome":  t.get("outcome"),
            "duration": round(t.get("duration", 0), 3),
            "error":    _get_test_error(t) if t.get("outcome") != "passed" else "",
        }
        for t in report.get("tests", [])
    ]

    # ── Update leaderboard (GREATEST keeps highest score) ─────────────────────
    upsert_score(
        username=user["sub"],
        section=body.section,
        score=score,
        passed_tests=passed,
        total_tests=total,
        time_taken=round(duration, 2),
        full_name=user.get("full_name", ""),
        college=user.get("college", ""),
        team=user.get("team", ""),
    )

    # ── Anti-cheat report and risk score ──────────────────────────────────────
    risk_score, risk_flags, risk_level = build_risk(
        test_hash_valid=test_hash_valid,
        tests_collected=collected,
        tests_skipped=skipped,
        suspicious_occurrences=suspicious_occurrences,
        duplicate_detected=duplicate_detected,
        submission_rate_last_10min=submission_rate_last_10min,
        runtime_seconds=duration,
        paste_attempts=copy_metrics.get("paste_attempts", 0),
        large_injection_events=copy_metrics.get("large_injection_events", 0),
        typing_anomaly_detected=copy_metrics.get("typing_anomaly_detected", False),
        copy_risk_score=copy_metrics.get("copy_risk_score", 0),
        expected_total_tests=57,
    )

    if test_hash_mismatches:
        risk_flags.append(f"Mismatched tests: {', '.join(test_hash_mismatches)}")
    if duplicate_with:
        risk_flags.append(f"Duplicate with team: {duplicate_with}")

    team_id = user.get("team") or user.get("sub") or "unknown"
    anti_cheat_report = {
        "username": user.get("sub", ""),
        "team_id": team_id,
        "section": body.section,
        "submission_hash": submission_hash,
        "tests_collected": collected,
        "tests_passed": passed,
        "tests_failed": failed,
        "tests_skipped": skipped,
        "runtime_seconds": round(duration, 4),
        "cpu_usage_percent": cpu_usage_percent,
        "memory_usage_mb": memory_usage_mb,
        "suspicious_imports": suspicious_imports,
        "test_hash_valid": test_hash_valid,
        "duplicate_detected": duplicate_detected,
        "submission_rate_last_10min": submission_rate_last_10min,
        "paste_attempts": copy_metrics.get("paste_attempts", 0),
        "large_injection_events": copy_metrics.get("large_injection_events", 0),
        "typing_anomaly_detected": copy_metrics.get("typing_anomaly_detected", False),
        "copy_risk_score": copy_metrics.get("copy_risk_score", 0),
        "average_typing_speed_cps": copy_metrics.get("average_typing_speed_cps", 0.0),
        "tab_switches": copy_metrics.get("tab_switches", 0),
        "window_blur_seconds": copy_metrics.get("window_blur_seconds", 0.0),
        "risk_score": risk_score,
        "risk_level": risk_level,
        "risk_flags": risk_flags,
        "raw_pytest_output": f"{raw_stdout}\n{raw_stderr}".strip(),
    }
    anti_cheat_report_id = save_anti_cheat_report(anti_cheat_report)

    return {
        "section":          body.section,
        "passed":           passed,
        "failed":           failed,
        "total":            total,
        "score":            score,
        "max_score":        max_pts,
        "duration_seconds": round(duration, 2),
        "results":          tests,
        "anti_cheat_report_id": anti_cheat_report_id,
        "anti_cheat": {
            "team_id": team_id,
            "submission_hash": submission_hash,
            "tests_collected": collected,
            "tests_passed": passed,
            "tests_failed": failed,
            "tests_skipped": skipped,
            "runtime_seconds": round(duration, 4),
            "cpu_usage_percent": cpu_usage_percent,
            "memory_usage_mb": memory_usage_mb,
            "suspicious_imports": suspicious_imports,
            "test_hash_valid": test_hash_valid,
            "duplicate_detected": duplicate_detected,
            "submission_rate_last_10min": submission_rate_last_10min,
            "paste_attempts": copy_metrics.get("paste_attempts", 0),
            "large_injection_events": copy_metrics.get("large_injection_events", 0),
            "typing_anomaly_detected": copy_metrics.get("typing_anomaly_detected", False),
            "copy_risk_score": copy_metrics.get("copy_risk_score", 0),
            "average_typing_speed_cps": copy_metrics.get("average_typing_speed_cps", 0.0),
            "tab_switches": copy_metrics.get("tab_switches", 0),
            "window_blur_seconds": copy_metrics.get("window_blur_seconds", 0.0),
            "risk_score": risk_score,
            "risk_flags": risk_flags,
            "risk_level": risk_level,
        },
        "note":             "Only your highest score per section is kept.",
    }

File: backend/routers/validator.js
'use strict';
/**
 * Validator router — clone a participant's GitHub repo and run
 * section pytest tests against it, then record score + anti-cheat report.
 */

const router  = require('express').Router();
const path    = require('path');
const fs      = require('fs');
const os      = require('os');
const { execFileSync, spawn } = require('child_process');
const config  = require('../core/config');
const db      = require('../core/database');
const { requireAuth } = require('../middleware/auth');
const { verifyTestIntegrity, scanSuspiciousImports, hashSourceTree } = require('../anti_cheat');
const { buildRisk } = require('../risk_engine');

const SECTION_POINTS = { 1: 100, 2: 100, 3: 100, 4: 120 };
const SECTION_TESTS  = {
  1: 'section1-multifile-debug/tests',
  2: 'section2-broken-recovery/tests',
  3: 'section3-memory-deadlock/tests',
  4: 'section4-logical-tracing/tests',
};

// Clone with depth=1 and a strict timeout; no interactive prompts
function gitClone(url, dest) {
  return new Promise((resolve, reject) => {
    const env = {
      ...process.env,
      GIT_TERMINAL_PROMPT: '0',
      GIT_ASKPASS: 'echo',
    };
    const proc = spawn('git', [
      'clone', '-c', 'credential.helper=',
      '--depth=1', '--no-tags', url, dest,
    ], { env });

    let stderr = '';
    proc.stderr.on('data', d => { stderr += d.toString(); });

    const timer = setTimeout(() => {
      proc.kill('SIGKILL');
      reject(Object.assign(new Error('Clone timed out after 90s'), { status: 408 }));
    }, 90_000);

    proc.on('close', code => {
      clearTimeout(timer);
      if (code === 0) resolve();
      else reject(Object.assign(new Error(`git clone failed: ${stderr.slice(0, 400).trim()}`), { status: 422 }));
    });
    proc.on('error', err => { clearTimeout(timer); reject(err); });
  });
}

function runPytest(testPath, reportPath, cwd) {
  return new Promise((resolve, reject) => {
    const timeoutMs = (config.PYTEST_TIMEOUT + 30) * 1000;
    const proc = spawn('python3', [
      '-m', 'pytest', testPath,
      '--json-report', `--json-report-file=${reportPath}`,
      '--tb=no', '-q', `--timeout=${config.PYTEST_TIMEOUT}`,
    ], { cwd });

    let killed = false;
    const timer = setTimeout(() => {
      killed = true; proc.kill('SIGKILL');
      reject(Object.assign(new Error('pytest timed out'), { status: 408 }));
    }, timeoutMs);

    proc.on('close', () => { clearTimeout(timer); if (!killed) resolve(); });
    proc.on('error', err => { clearTimeout(timer); reject(err); });
  });
}

// POST /api/validate/editor-activity
router.post('/editor-activity', requireAuth, async (req, res, next) => {
  try {
    const { event, team_id, timestamp, editor_length_before=0, editor_length_after=0,
            chars_delta=0, delta_ms=0, typing_speed_cps=0, flags=[], details={} } = req.body;
    const teamId = team_id || req.user.team || req.user.sub || 'unknown';
    const id = await db.logEditorActivity({
      username: req.user.sub || '', teamId, event,
      editorLengthBefore: editor_length_before, editorLengthAfter: editor_length_after,
      charsDelta: chars_delta, deltaMs: delta_ms, typingSpeedCps: typing_speed_cps,
      metadata: { timestamp, flags, details },
    });
    res.json({ ok: true, id });
  } catch (err) { next(err); }
});

// POST /api/validate/section
router.post('/section', requireAuth, async (req, res, next) => {
  const { section, github_url: githubUrl } = req.body;

  if (!SECTION_TESTS[section])
    return res.status(422).json({ detail: 'section must be 1–4' });
  if (!githubUrl)
    return res.status(422).json({ detail: 'github_url is required' });

  // ── Timer check ──────────────────────────────────────────────────────────
  const timer = await db.getSectionTimer(section);
  if (timer && timer.submissions_locked)
    return res.status(423).json({ detail: `Section ${section} submissions are locked by the administrator.` });
  if (timer && !timer.is_active)
    return res.status(403).json({ detail: `Section ${section} has not been started yet by the administrator.` });
  if (timer && timer.is_active && timer.start_time) {
    const start    = new Date(timer.start_time);
    const pausedAt = timer.paused_at ? new Date(timer.paused_at) : null;
    const elapsed  = timer.elapsed_seconds || 0;
    const totalSecs = timer.duration_minutes * 60;
    const now = new Date();

    let remaining;
    if (pausedAt) {
      remaining = totalSecs - elapsed;
    } else {
      const runningSecs = Math.floor((now - start) / 1000);
      remaining = totalSecs - elapsed - runningSecs;
    }
    if (remaining <= 0)
      return res.status(403).json({ detail: `Section ${section} time has expired. No more submissions accepted.` });
  }

  const teamId = req.user.team || req.user.sub || 'unknown';

  // ── GitHub username enforcement ────────────────────────────────────────────
  const student = await db.getStudent(req.user.sub);
  const regGhUser = student ? (student.github_username || '').trim().toLowerCase() : '';
  if (!regGhUser)
    return res.status(422).json({ detail: 'You must set a GitHub username on the Dashboard before validating.' });

  // Extract owner from the submitted URL: https://github.com/<owner>/<repo>
  const ghUrlMatch = githubUrl.match(/^https:\/\/github\.com\/([^\/]+)\/([^\/]+?)(\.git)?$/i);
  if (!ghUrlMatch)
    return res.status(422).json({ detail: 'Invalid GitHub URL. Use the format: https://github.com/owner/repo' });
  const repoOwner = ghUrlMatch[1].toLowerCase();
  if (repoOwner !== regGhUser)
    return res.status(403).json({
      detail: `Repository owner "${ghUrlMatch[1]}" does not match your registered GitHub username "${student.github_username}". You may only submit from your own account.`,
    });

  let tmpdir = null;
  let reportPath = null;

  try {
    tmpdir = fs.mkdtempSync(path.join(os.tmpdir(), 'dc_validate_'));

    // ── Clone ───────────────────────────────────────────────────────────────────────
    await gitClone(githubUrl, tmpdir);


    // ── Verify tests exist ─────────────────────────────────────────────────
    const testPath = path.join(tmpdir, SECTION_TESTS[section]);
    if (!fs.existsSync(testPath))
      return res.status(422).json({
        detail: `Tests for section ${section} not found in repo. Expected: ${SECTION_TESTS[section]}`,
      });

    // ── Anti-cheat pre-checks ──────────────────────────────────────────────
    const { valid: testHashValid, mismatched: testHashMismatches } =
      verifyTestIntegrity(config.REPO_ROOT, tmpdir);
    const { imports: suspiciousImports, occurrences: suspiciousOccurrences } =
      scanSuspiciousImports(tmpdir);
    const submissionHash = hashSourceTree(tmpdir);

    const duplicateRow = await db.findDuplicateSubmissionHash(submissionHash, teamId);
    const duplicateDetected = duplicateRow !== null;
    const submissionRateLast10min = await db.countRecentSubmissions(teamId, 10);
    const copyMetrics = await db.getEditorActivityMetrics(teamId, 10);

    // ── Run pytest ─────────────────────────────────────────────────────────
    reportPath = path.join(os.tmpdir(), `dc_report_${Date.now()}_${Math.random().toString(36).slice(2)}.json`);
    const startTs = Date.now();
    try {
      await runPytest(testPath, reportPath, tmpdir);
    } catch (err) {
      if (err.status === 408) return res.status(408).json({ detail: 'Test run timed out' });
      // pytest exits non-zero when tests fail — that's fine; report will still exist
    }
    const duration = (Date.now() - startTs) / 1000;

    // ── Parse report ───────────────────────────────────────────────────────
    let report;
    try {
      report = JSON.parse(fs.readFileSync(reportPath, 'utf8'));
    } catch {
      return res.status(500).json({ detail: 'Could not parse pytest report' });
    }

    const summary   = report.summary || {};
    const passed    = summary.passed  || 0;
    const failed    = (summary.failed || 0) + (summary.error || 0);
    const total     = summary.total   || (passed + failed);
    const skipped   = summary.skipped || 0;
    const collected = summary.collected || total;

    const maxPts = SECTION_POINTS[section];
    const score  = total ? Math.round((passed / total) * maxPts * 100) / 100 : 0;

    const rawOutput = report.tests
      ? report.tests.map(t => `${t.nodeid}: ${t.outcome}`).join('\n')
      : '';

    // ── Risk scoring ───────────────────────────────────────────────────────
    const { score: riskScore, flags: riskFlags, level: riskLevel } = buildRisk({
      testHashValid,
      testsCollected:           collected,
      testsSkipped:             skipped,
      suspiciousOccurrences,
      duplicateDetected,
      submissionRateLast10min,
      runtimeSeconds:           duration,
      pasteAttempts:            copyMetrics.paste_attempts,
      largeInjectionEvents:     copyMetrics.large_injection_events,
      typingAnomalyDetected:    copyMetrics.typing_anomaly_detected,
      copyRiskScore:            copyMetrics.copy_risk_score,
    });

    // ── Save anti-cheat report ─────────────────────────────────────────────
    const student = await db.getStudent(req.user.sub);
    const antiCheatId = await db.saveAntiCheatReport({
      username:                  req.user.sub,
      team_id:                   teamId,
      section,
      submission_hash:           submissionHash,
      tests_collected:           collected,
      tests_passed:              passed,
      tests_failed:              failed,
      tests_skipped:             skipped,
      runtime_seconds:           duration,
      cpu_usage_percent:         0,
      memory_usage_mb:           0,
      suspicious_imports:        suspiciousImports,
      test_hash_valid:           testHashValid,
      duplicate_detected:        duplicateDetected,
      submission_rate_last_10min: submissionRateLast10min,
      paste_attempts:            copyMetrics.paste_attempts,
      large_injection_events:    copyMetrics.large_injection_events,
      typing_anomaly_detected:   copyMetrics.typing_anomaly_detected,
      copy_risk_score:           copyMetrics.copy_risk_score,
      average_typing_speed_cps:  copyMetrics.average_typing_speed_cps,
      tab_switches:              copyMetrics.tab_switches,
      window_blur_seconds:       copyMetrics.window_blur_seconds,
      risk_score:                riskScore,
      risk_level:                riskLevel,
      risk_flags:                riskFlags,
      raw_pytest_output:         rawOutput,
    });

    // ── Upsert leaderboard ─────────────────────────────────────────────────
    await db.upsertScore({
      username:       req.user.sub,
      section,
      score,
      passedTests:    passed,
      totalTests:     total,
      timeTaken:      duration,
      fullName:       req.user.full_name || '',
      college:        req.user.college   || '',
      team:           req.user.team      || '',
      teamNo:         '',
      teamName:       req.user.team || '',
      commitNo:       '',
      githubRepoLink: githubUrl,
    });

    // ── Build per-test details ─────────────────────────────────────────────
    const tests = (report.tests || []).map(t => {
      let error = '';
      for (const phase of ['call', 'setup', 'teardown']) {
        const pd = t[phase] || {};
        if (['failed', 'error'].includes(pd.outcome)) {
          error = typeof pd.longrepr === 'object' ? (pd.longrepr?.repr || '') : (pd.longrepr || '');
          break;
        }
      }
      return {
        name:     (t.nodeid || '').split('::').pop(),
        outcome:  t.outcome,
        duration: Math.round((t.duration || 0) * 1000) / 1000,
        error:    t.outcome !== 'passed' ? error.trim() : '',
      };
    });

    res.json({
      section,
      score,
      max_score:          maxPts,
      passed,
      failed,
      total,
      skipped,
      duration_seconds:   Math.round(duration * 100) / 100,
      results:            tests,
      anti_cheat_report_id: antiCheatId,
      risk_level:         riskLevel,
      risk_score:         riskScore,
      risk_flags:         riskFlags,
    });

  } catch (err) {
    if (err.status) return res.status(err.status).json({ detail: err.message });
    next(err);
  } finally {
    // Always clean up temp files
    if (reportPath && fs.existsSync(reportPath)) {
      try { fs.unlinkSync(reportPath); } catch {}
    }
    if (tmpdir && fs.existsSync(tmpdir)) {
      try { fs.rmSync(tmpdir, { recursive: true, force: true }); } catch {}
    }
  }
});

module.exports = router;
